core/src/test/java/org/infinispan/container/SimpleDataContainerTest.java

```diff
@@ -283,4 +291,41 @@ public void testGetDuringKeySetLoop() {
 
       assertEquals(10, i);
    }
+
+   public void testEntrySetStreamWithExpiredEntries() {
+      dc.put("k1", "v1", new EmbeddedMetadata.Builder().lifespan(100, TimeUnit.MILLISECONDS).build());
+      dc.put("k2", "v2", new EmbeddedMetadata.Builder().build());
+      dc.put("k3", "v3", new EmbeddedMetadata.Builder().maxIdle(200, TimeUnit.MILLISECONDS).build());
+
+      Set<Map.Entry<String, String>> expected = new HashSet<>();
+      Map.Entry<String, String> k1 = CoreImmutables.immutableInternalCacheEntry(dc.get("k1"));
+      expected.add(k1);
+      expected.add(CoreImmutables.immutableInternalCacheEntry(dc.get("k2")));
+      Map.Entry<String, String> k3 = CoreImmutables.immutableInternalCacheEntry(dc.get("k3"));
+      expected.add(k3);
+
+      List<Map.Entry<String, String>> results = Arrays.asList(dc.entrySet().stream().toArray(Map.Entry[]::new));
+      assertEquals(3, results.size());
+      assertArrayAndSetContainSame(expected, results);
+
+      timeService.advance(101);
+
+      results = Arrays.asList(dc.entrySet().stream().toArray(Map.Entry[]::new));
+      assertEquals(2, results.size());
+      expected.remove(k1);
+      assertArrayAndSetContainSame(expected, results);
+
+      timeService.advance(100);
+
+      results = Arrays.asList(dc.entrySet().stream().toArray(Map.Entry[]::new));
+      assertEquals(1, results.size());
```

Comment:
Iterator doesn't touch them. Otherwise it totally throws off the eviction ordering. We can discuss this on dev list if we want to change this maybe?